Create Recoverable Wallet Contracts

mhfen opened the pull request:
### What's this PR do?
- Creates a recoverable contract
- Creates a basic wallet contract that is recoverable

Still need to write tests, but I thought I would put this up for review while I do that.

### Where should the reviewer start?
Recoverable.sol has all of the important work.

### Any background context you want to provide?
Recoverable is a recoverable contract with owner set to the original sender and required recovery time set by the original sender. The owner can set a recoverer at any time. The owner can transfer ownership at any time. The recoverer can initiate a recovery at any time. The recoverer must wait for the required recovery time in order to complete the recovery by making themselves owner. The owner can block the recovery if they do so within the required recovery time.

### Does this add new dependencies? Is there an installation procedure? (bundle, bower, npm, etc)
No.

mhfen committed:
Create a recoverable contract

- Contract has an owner
- Contract owner sets a recovery address
- Recovery address can inititiate a recovery
- Once recovery time has lapsed, recovery address can set themselves to owner

mhfen committed:
Create basic wallet contract on top of Recoverable

dadeg commented on contracts/Recoverable.sol:
```diff
@@ -1,5 +1,133 @@
+  /**
+   * The transfer mechanism
+   */
+  function _transferOwnership(address _newOwner) internal {
```
We can use the openzepp "ownable" contract for this by inheriting it. Claimable is probably a better one, too. It allows you to transfer ownership but the new owner must claim it. This stops you from accidentally setting ownership on an account that doesn't exist

dadeg commented on contracts/Recoverable.sol:
```diff
@@ -1,5 +1,133 @@
+contract Recoverable {
+  address public owner_;
+  address private recoverer_;
+  uint256 private recoveryTimeRequred_;
```
Typo in required

dadeg commented on contracts/Recoverable.sol:
```diff
@@ -1,5 +1,133 @@
+  );
+  event RecoveryStarted(
+    address indexed owner_,
+    address indexed heir,
```
Heir is misleading.. how about destination?

dadeg commented on contracts/Recoverable.sol:
```diff
@@ -1,5 +1,133 @@
+   * Check for whether recovery time has lapsed
+   */
+  function recoveryTimeLapsed() internal view returns (bool) {
+    return block.timestamp <= timeOfRecovery_ + recoveryTimeRequred_;
```
I believe this should be greater than. Just a "tiny" bug lol

dadeg commented on contracts/Recoverable.sol:
```diff
@@ -1,5 +1,133 @@
+    recoverer_ = _newRecoverer;
+  }
+
+  function recoverer() public view returns(address) {
```
You don't need these getters if you just make the property itself public. They are made automatically. Setters are NOT made automatically though

dadeg commented on contracts/Recoverable.sol:
```diff
@@ -1,5 +1,133 @@
+   */
+  function announceRecovery() public isOwner {
+    require(recoveryTimeLapsed());
+    emit RecoveryStarted(owner_, recoverer_, timeOfRecovery_);
```
It looks like maybe you meant to emit the event with the current timestamp? But the setting of time happens afterwards?

dadeg commented on contracts/Recoverable.sol:
```diff
@@ -1,5 +1,133 @@
+   */
+  function blockRecovery() public isOwner {
+    emit OwnerBlockedRecovery(owner_);
+    timeOfRecovery_ = 0;
```
How does this stop recovery?

mhfen replied:
Because `recoveryTimeLapsed` will never return true when this is set to zero.

dadeg commented on contracts/Recoverable.sol:
```diff
@@ -1,5 +1,133 @@
+   * Once reverty time has lapsed, a recoverer may transfer ownership
+   * to themselves
+   */
+  function recoverContract() public isRecoverer {
```
I think my preference is that the funds are sent to a new address instead of changing the owner. Firstly, the recoverer might not want to deal with being the owner. Secondly, the wallet might not be always so simple as having a single owner, or it might not have an owner at all. Thirdly, the wallet might be broken in some way that makes it impossible to do anything except recover it.

mhfen replied:
1. Whether the recoverer is set to owner or the funds are sent to their address, they are at the end of the day the owner until recovery by the original owner. Seems to me it wouldn't matter, and this is the truest implementation of a "recoverable" wallet. If the funds are sent somewhere else, the wallet isn't really recovered, it's abandoned.
2. I don't understand this one. We are creating this Recoverable wallet for someone to own. I suppose it could be extended for multiple owners in the future, but we don't need to do that here.
3. This doesn't change anything. The recoverer could then send the funds somewhere if the rest of the wallet is broken. We should probably make it unbreakable tho...

dadeg commented on contracts/RecoverableWallet.sol:
```diff
@@ -1,5 +1,23 @@
+/**
+ * A basic wallet contract that is recoverable by another party.
+ */
+contract RecoverableWallet is Recoverable {
```
I think the owner stuff can be on this contract. Since recovery will not have the notion of owner any more.

dadeg replied:
Now that I see this, it seems we could just inherit the openzepp wallet as well and not need to copy the code at all. The wallet class will just be an empty block

mhfen replied:
Owner stuff can't be in this contract since we need it to limit who can do what in Recoverable.

mhfen committed:
Address PR Feedback

- Fixed typo
- Fixed recoverer reference
- Fixed timeelapsed check

mhfen committed:
Add openzeppelin-solidity

Open source contract library

mhfen committed:
Replace owner functionality with base Ownable

OpenZep already has a lot of this functionality in Ownable, so using that as a base for Recoverable.

mhfen committed:
change transfer ownership to send funds

mhfen committed:
mistyped _sendTo

mhfen committed:
use proper reference for balance
